chat_v2_message_counter_sample/src/main/java/com/zendesk/chat_v2/sample/UnreadMessageCounter.java

```diff
@@ -0,0 +1,75 @@
+package com.zendesk.chat_v2.sample;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import zendesk.chat.Chat;
+import zendesk.chat.ChatLog;
+import zendesk.chat.ChatState;
+import zendesk.chat.ObservationScope;
+import zendesk.chat.Observer;
+
+class UnreadMessageCounter {
+
+    private final UnreadMessageListener unreadMessageListener;
+
+    private boolean shouldCount;
+    private String lastChatLogId;
+    private String lastReadChatLogId;
+    private List<ChatLog> unreadChatLogs = new ArrayList<>();
+    private ObservationScope observationScope = new ObservationScope();
+
+    public UnreadMessageCounter(UnreadMessageListener unreadMessageListener) {
+        this.unreadMessageListener = unreadMessageListener;
+
+        Chat.INSTANCE.providers().chatProvider().observeChatState(observationScope, new Observer<ChatState>() {
+            @Override
+            public void update(ChatState chatState) {
+                if (chatState != null && !chatState.getChatLogs().isEmpty()) {
+                    if (shouldCount && lastReadChatLogId !=  null) {
+                        updateCounter(chatState.getChatLogs());
+                    }
+                    lastChatLogId = chatState.getChatLogs().get(
+                            chatState.getChatLogs().size() - 1
+                    ).getId();
+                }
+            }
+        });
+    }
+
+
+    // Determines whether or not the chat websocket should be re-connected to.
+    // A connected, open websocket will disable push notifications and receive messages as normal.
+    public void setShouldCount(boolean shouldCount) {
+        if (shouldCount) {
+            lastReadChatLogId = lastChatLogId;
+            Chat.INSTANCE.providers().connectionProvider().connect();
+        }
+        this.shouldCount = shouldCount;
+    }
+
+    public void stopCounter() {
+        unreadChatLogs = new ArrayList<>();
+        lastReadChatLogId = null;
+        setShouldCount(false);
+        unreadMessageListener.onUnreadUpdated(0);
+        observationScope.cancel();
```

Comment:
If we cancel this here, do we need to make a new one if we wish to start observing again in the future?

Reply:
Hmm yes, good point.